clang/lib/AST/DeclTemplate.cpp

```diff
@@ -51,14 +51,15 @@ DefaultTemplateArgumentContainsUnexpandedPack(const TemplateParam &P) {
          P.getDefaultArgument().getArgument().containsUnexpandedParameterPack();
 }
 
-TemplateParameterList::TemplateParameterList(const ASTContext& C,
+TemplateParameterList::TemplateParameterList(const ASTContext &C,
                                              SourceLocation TemplateLoc,
                                              SourceLocation LAngleLoc,
                                              ArrayRef<NamedDecl *> Params,
                                              SourceLocation RAngleLoc,
                                              Expr *RequiresClause)
-    : TemplateLoc(TemplateLoc), LAngleLoc(LAngleLoc), RAngleLoc(RAngleLoc),
-      NumParams(Params.size()), ContainsUnexpandedParameterPack(false),
+    : InjectedArgs(&C), TemplateLoc(TemplateLoc), LAngleLoc(LAngleLoc),
```

Comment:
This trickery with storing the ASTContext seems a little silly. I would vastly prefer one of the following:

1- We just store the AST Context reference in this type.
2- We make `getInjectedTemplateArgs` take an `ASTContext` reference.
3- We take a collection in `getInjectedTemplateArgs` and just fill it.
4- We DON'T have this function (or do #2), and keep the `ASTContext` version of this function to do the work.

Reply:
> This trickery with storing the ASTContext seems a little silly.

@erichkeane Is there any particular reason why? We already do something similar in e.g. `Redeclarable::DeclLink`. The injected template arguments are lazily allocated & cached, so we don't need the pointer to `ASTContext` once the injected template arguments have been allocated. Since we _can_ use a `PointerUnion` to store both values, I think adding yet another pointer to `TemplateParameterList` would be rather unjustified.

Reply:
Honestly, I strongly dislike it there as well.  It seems like quite a hacky way of getting our intended behavior here that is jarring/unreadable every time I see it.  I also think that there is some level of awkwardness to doing an allocation in a function whose type doesn't really look like it should be able to allocate.

Reply:
> I also think that there is some level of awkwardness to doing an allocation in a function whose type doesn't really look like it should be able to allocate.

Since it's an accessor for lazily allocated data, it has to allocate if the data hasn't been allocated yet :)

Reply:
Yep, i understand that.  Currying is always pretty awkward in languages/programs that don't otherwise lazily evaluate :)  It seems to me if the callers to this all have access to ASTContext, we could more easily just require it as a caller-passed arg.

Reply:
@erichkeane What about the `getInjectedTemplateArgs` functions declared in `RedeclarableTemplateDecl`, `ClassTemplatePartialSpecializationDecl`, and `VarTemplatePartialSpecializationDecl`? Sure, you _could_ just call `getASTContext` and pass the result, but that involves walking up the parent `DeclContexts` until we hit the `TranslationUnitDecl`.

Even if we hada `ASTContext` parameter, we still end up with a function that "doesn't really look like it should be able to allocate" & mutable AST state (the same as the current implementation). If you _really_ do feel strongly about this then alternatives could be considered, but I would overwhelmingly prefer the current approach with additional comments explaining why a pointer to `ASTContext` is stored.

Reply:
My hope would be that the callers there could pass on their AST Context, I see the uses in this review are either already have an ASTContext or access to Sema.  Could you do that (add it as a parm to the `getInjectedTemplateArgs`, and see how rough it is for the callers to get the ASTContext), and put it somewhere I can take a look?  If it ends up getting too rough in some places, I think i could 'disagree and commit' to this approach.

Reply:
> > Could you do that (add it as a parm to the `getInjectedTemplateArgs`, and see how rough it is for the callers to get the ASTContext), and put it somewhere I can take a look?
> 
> @erichkeane See [3fd8326](https://github.com/llvm/llvm-project/commit/3fd8326e222877bce474010481a459b5db0c8be2)
> 
> This _will_ break tools that use `getInjectedTemplateArgs` though (e.g. the tool I'm working on [MrDocs] uses `getInjectedTemplateArgs`).

Thanks for doing that!  I'll note that looks way nicer IMO.  I'd hope that downstream users would have access to the ASTContext/Sema in a way that they could use that?  Is it REALLY troublesome/not really possible, or is it just a code change?

Reply:
> Is it REALLY troublesome/not really possible, or is it just a code change?

Fixing existing code should be trivial, I just really dislike "getters" with parameters :P Ultimately, the "right thing" here is very subjective... but I'll cherry-pick [3fd8326](https://github.com/llvm/llvm-project/commit/3fd8326e222877bce474010481a459b5db0c8be2) onto this branch & merge, absent any objections

Reply:
>Fixing existing code should be trivial, I just really dislike "getters" with parameters :P 
I'm not THRILLED with it either, though we do this with `ASTContext` all over the place in the codebase to mean 'this probably needs to allocate'.  But thank you for your patience/flexibility here.